cpp/arcticdb/util/reliable_storage_lock.tpp

```diff
@@ -0,0 +1,233 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/util/reliable_storage_lock.hpp>
+
+#include <arcticdb/storage/store.hpp>
+#include <arcticdb/entity/types.hpp>
+#include <arcticdb/util/constants.hpp>
+#include <arcticdb/util/configs_map.hpp>
+
+#include <mutex>
+#include <folly/experimental/FunctionScheduler.h>
+
+namespace arcticdb {
+
+namespace lock {
+
+const auto SEPARATOR = '*';
+const auto EXTENDS_PER_TIMEOUT = 5u;
+const auto KEEP_LAST_N_LOCKS = 5u;
+
+StreamDescriptor lock_stream_descriptor(const StreamId &stream_id) {
+    return StreamDescriptor{stream_descriptor(
+            stream_id,
+            stream::RowCountIndex(),
+            {scalar_field(DataType::INT64, "expiration")})};
+}
+
+SegmentInMemory lock_segment(const StreamId &name, timestamp expiration) {
+    SegmentInMemory output{lock_stream_descriptor(name)};
+    output.set_scalar(0, expiration);
+    output.end_row();
+    return output;
+}
+
+template <class ClockType>
+ReliableStorageLock<ClockType>::ReliableStorageLock(const std::string &base_name, const std::shared_ptr<Store> store, timestamp timeout) :
+    base_name_(base_name), store_(store), timeout_(timeout) {
+    auto s3_timeout = ConfigsMap::instance()->get_int("S3Storage.RequestTimeoutMs", 200000) * ONE_MILLISECOND;
```

Comment:
Yeah we'll want to use it for azure etc which already has similar operations, it would be better to pass it in as a policy from the storage where we just expect a function that is void() that does whatever we need